Dev

monicah79 opened the pull request:
Only admins can add laptops and this is the admins password admin_password:69420lesgooo
- followed the layout of [the provided design](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign)
- Selected a theme which is laptop eccommerce
- The user logs in to the website, only by typing the username
- In the navigation panel, the user can see links to:
"Reserve" form.
"My reservations".
"Add laptops/items that you selected as a theme" (the link is visible to everybody).
"Delete laptops/items that you selected as a theme" (the link is visible to everybody).
- On the main page, the user can see a [list](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005577) before signing up bt cant make a reservation
- When the user selects a specific item, they can see the [details page](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005579) with its full description (skip the "Rotate image" button).
- WheMake the app responsive, creating both mobile and desktop versionsn the user clicks the "Add item" link in the navigation panel they can see a form for adding a new item.
- To reserve an appointment, the user has to select a date and city (username and selected item are autofilled).
Use the design based on the ["Book a vespa test-ride"](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005583) and add all necessary inputs.
The user can also access the "Reserve" page from the navigation panel. In that case only username is autofilled.
- When the user clicks the "My reservations" link in the navigation panel they can see a list of their reservations (with information about item name, date and city).
- we added full documentation for your API.
- Implemented proper user authentication from the front-end to the server.
- Make sure that the "Add item" and "Delete item" links are accessible only by users who are admins.

monicah79 committed:
Capstone setup devise

monicah79 committed:
Updated routes

monicah79 committed:
Edit

monicah79 committed:
Removed .env file from repository

monicah79 committed:
Removed .env file from repository

monicah79 committed:
Merge pull request #7 from monicah79/Capstone-setup

Capstone setup devise

shasherazi committed:
Add linux platform in gemfile

monicah79 committed:
Jwt with devise

monicah79 committed:
Merge pull request #18 from monicah79/jwt-devise

Jwt with devise

shasherazi committed:
Create reservations model

shasherazi committed:
Add reservations controller

shasherazi committed:
Add request specs for Reservations

shasherazi committed:
Add model specs for Reservation

shasherazi committed:
Linters fix

shasherazi committed:
Linters fix

fuadmoin committed:
Create model for laptop.

fuadmoin committed:
Add final touches.

shasherazi committed:
Merge pull request #19 from monicah79/setup-reservations

Setup reservations

fuadmoin committed:
Merge pull request #20 from monicah79/features/laptop-model

Features/laptop model

shasherazi committed:
Add show and delete methods for reservations

shasherazi committed:
Merge pull request #21 from monicah79/add-show-and-delete-reservation-methods

Add show and delete methods for reservations

fuadmoin committed:
Add controller for laptop model.

fuadmoin committed:
Merge branch 'dev' into features/laptop-controller

fuadmoin committed:
Fix linter errors.

fuadmoin committed:
Fix linter errors.

fuadmoin committed:
Merge pull request #22 from monicah79/features/laptop-controller

Add controller for laptop model.

monicah79 committed:
Laptop reservation model

monicah79 committed:
Merge pull request #25 from monicah79/laptop-reservation-model

Laptop reservation model

fuadmoin committed:
Add unit tests for the Laptop model.

monicah79 committed:
Laptop reservation controller

monicah79 committed:
Updated laptop reservation

monicah79 committed:
Updated readme

monicah79 committed:
Updated license

monicah79 committed:
updated laptop reservation

monicah79 committed:
Updated laptop reservations

fuadmoin committed:
Update credentials.yml

fuadmoin committed:
Merge pull request #29 from monicah79/features/laptop-tests

Add unit tests for the Laptop model.

fuadmoin committed:
Add tests for the laptop controller.

monicah79 committed:
Requested chages

monicah79 committed:
Requested chages

monicah79 committed:
Merge pull request #30 from monicah79/laptop-reservation-controller

Laptop reservation controller

fuadmoin committed:
Merge pull request #31 from monicah79/features/laptop-tests

Add tests for the laptop controller.

shasherazi committed:
Fix the creation of laptop reservation

Please note the following (probably breaking) changes

- There is only the `show` route for `reservations` now. This is to make only the needed methods available in the API. One should not be creating reservations manually, they will already be created using `laptop_reservations`
- Please update the tests accordingly.

shasherazi committed:
Merge pull request #32 from monicah79/fix-laptop-reservation

Fix the creation of laptop reservation

fuadmoin committed:
Set up rswag api documentaion.

fuadmoin committed:
Add picture to the laptop table.

I added picture to the laptop table, and added validations for cpu and picture columns in the Laptop table.

fuadmoin committed:
Add documentaion for the project api.

fuadmoin committed:
Merge branch 'dev' into api-documentation

monicah79 committed:
Reservation spec

monicah79 committed:
Fixed linters

monicah79 committed:
Update

monicah79 committed:
Update

monicah79 committed:
Update

monicah79 committed:
Merge pull request #34 from monicah79/laptop-reservations-spec

Reservation spec

monicah79 committed:
Can can

shasherazi committed:
Allow user to signup as admin

Users can signup to be an admin, but only if they know a super secret password

shasherazi committed:
Merge pull request #37 from monicah79/cancan

Can can

shasherazi committed:
Merge branch 'api-documentation' into dev

fuadmoin committed:
Update api documentation.

fuadmoin committed:
Merge branch 'dev' into api-documentation

fuadmoin committed:
Update api documentation.

fuadmoin committed:
Remove unecessary files.

shasherazi committed:
Fix linters

shasherazi committed:
Merge pull request #38 from monicah79/api-documentation

Api documentation

imranhossainemi committed:
Update .gitignore

imranhossainemi committed:
Update laptops_controller.rb

imranhossainemi committed:
Update ability.rb

imranhossainemi committed:
Merge pull request #39 from monicah79/fix_public_laptop_view

Fix public laptop view

monicah79 committed:
Updated readme

monicah79 committed:
Merge pull request #40 from monicah79/update-readme

Updated readme

fuadmoin committed:
Modify the cre creation of reservation.

monicah79 committed:
Update README.md

Meltrust commented on README.md:
```diff
@@ -1 +1,164 @@
+- cd Final-capstone-backend
+- bundle install
+- rails db: create
+- rails db: migrate
```
- [x] When I try to create the database according to your instructions, this happens:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/b6b48fff-bad5-445c-a71d-8895e8e3f36c)

Kindly, include instructions in your readme about how to solve this issue:

```
1. Remove config/master.key and config/credentials.yml.enc if they exist.
2. Run `rails secret`.  Copy the key.
3. Run EDITOR="code --wait" bin/rails credentials:edit
4. In the editor that opens, add this:  devise_jwt_secret_key: <the key you copied in step 2>
5. Save the file and close the editor.  New master.key, credentials.yml.enc files will be generated, and the key will be stored in `Rails.application.credentials.devise_jwt_secret_key`.
```

That way, all your new users will be able to run your project properly. 👍

Meltrust commented on README.md:
```diff
@@ -1 +1,164 @@
-# Final-capstone-backend
\ No newline at end of file
+# Final-capstone-backend
```
# Frontend review

- [x] When  I start the app, I get an infinite requests to the server loop:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/6efce4b9-5a5f-415f-99d8-168ebd12fb3d)

Also, the rails console gets crazy trying to get "laptops", hundreds of times.

## Tips and potential fixes

This is the problem:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/79f0f190-61a4-4695-8901-21edc4908730)

Since your app is shipping with zero laptops, that `useEffect` will fire infinitely as long as there isn't at least one laptop. 

This also means all your new users will run into this.

You could:

1. Include some laptops (at least 1) in your seeds.rb file. (Recommended)
2. Hardcode at least one laptop in the frontend.
3. Debug that useEffect so it only runs `once` regardless of the case. (Recommended too).

Kindly, fix this requests loop so your users (and the server) have a better experience and your app runs as intended. 💪 

- [x] When I create a reservation, I immediately get redirected to "my reservations", but this happens:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/f353717b-4710-408f-ad50-1cbf113c0fdb)

When I reload the page, it works:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/ab7506cb-763b-430d-b95d-6f98121c6029)


This is because the reservation worked, and was recorded in the database, but the frontend variables didn't have the chance to update with the new data, and as a result, I get a blank page.

Kindly fix this bug so the app works as intended. 👍 

- [x]  In App.jsx please, kindly refactor the main `App` component to ES6 arrow function, so your code complies with the latest standards.

Refactoring is easy:

```javascript
const App = () => {
  // ...
}
```

## Explanation

The `function` keyword has been soft deprecated in almost all the software industry in favor of ES6 arrow functions. This means, almost nobody uses it anymore, so much that it's considered a non-best practice.

- [x] OPTIONAL: Refactor the rest of the functions (I found 7)

You can do a search in Vscode for the "function" word, and correct as needed.  Shouldn't take you more than 5-10 min so don't worry:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/fa8e809b-fd6e-400e-9b19-e5f61339457e)


That way, your code will meet the latest standards. 💪

- [x] Kindly, remove the 2 console.log statements found in `reservationSlice.js`:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/ccf09455-fb57-4865-8008-a8bad3069b58)

That way, your repo's cleanliness will be improved. 👍

Meltrust reviewed (changes requested):
Hi @monicah79,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  Nice code organization ✔️ 
-  API is working well✔️ 
-  Good readme ✔️ 




#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

![You can do it](https://media.giphy.com/media/FemTGpEZeK27M5fKL4/giphy.gif)

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

monicah79 committed:
Requested changes

monicah79 committed:
Added some files in seed file

monicah79 committed:
Update seed file

monicah79 committed:
Update seed file

Meltrust commented on db/seeds.rb:
```diff
@@ -0,0 +1,20 @@
+            laptop_id: laptop.id,
+            quantity: 1,
+            city: 'New York'
+          }
```
- [x] The seeds file is not working:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/e0047e56-144e-4683-a19b-35bf388be6a7)

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/510268aa-ef91-4fdc-836a-4d9b1c27cb82)

A working seeds.rb file that creates a user and a laptop is like this:

```ruby
# create a user
user = User.create!(email: 'user@example.com', password: 'password')

# create a laptop

laptop = Laptop.create!(name: 'MacBook Pro', price: 1299.99, cpu: 'Intel Core i7', memory: 16, picture: 'https', storage: 512,
                        user_id: 1)

puts "Created #{user}"
puts "Created #{laptop}"
```

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/2e349a50-047f-47e7-aac8-c6aa60a9c506)

After that, run `rails db:seed` and it will work:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/ab71587b-41f4-4c89-b737-c97557f09628)

In the frontend, the laptop will now show:

![image](https://github.com/monicah79/Final-capstone-backend/assets/57421823/2fc2aed1-67eb-4d4e-9f79-d93f7d200afd)


Kindly, go ahead and correct your seeds file. 💪

Meltrust reviewed (changes requested):
Hi @monicah79,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Functions are ES6✔️
-  Backend readme instructions are improved ✔️ 
-  Frontend infinite loop has been fixed✔️ 
-  Reservations creation bug has been fixed ✔️ 
-  console.logs have been removed ✔️ 



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

![You can do it](https://media.giphy.com/media/FemTGpEZeK27M5fKL4/giphy.gif)

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

fuadmoin committed:
Fix deletion of reservation.

Now when a laptop is deleted all reservations realted to it will also be deleted.

fuadmoin committed:
Merge branch 'final-touch' into dev
Update deletion of reservation.

fuadmoin committed:
Fix the seed file.

Meltrust reviewed (approved):
#### Hi @monicah79,


# **Wow, you did it 🎉**
![Brilliant](https://media.giphy.com/media/d31w24psGYeekCZy/giphy.gif)
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
![applause](https://media.giphy.com/media/srg19CG0cKMuI/giphy.gif)

### To highlight

- NIce job✔️ 



Cheers and Happy coding!👏👏👏


------

imranhossainemi committed:
Update README.md

imranhossainemi committed:
Update README.md

imranhossainemi committed:
Update README.md

imranhossainemi committed:
Create kanbanBoardImg.png

imranhossainemi committed:
Update README.md

imranhossainemi committed:
Update README.md
